### ISSUE TITLE
`TypeError` when initializing HMC for `ParamMixture` model

### ISSUE THREAD

**Issue by tscholak** [blei-lab/edward#610]
Consider the forward model
```python
pi = Dirichlet(pi_alpha)
mu = Normal(0., mu_sigma, sample_shape=[K])
sigmasq = InverseGamma(sigmasq_alpha, sigmasq_beta, sample_shape=[K])
x = ParamMixture(pi,
                 {'mu': mu, 'sigma': tf.sqrt(sigmasq)},
                 Normal,
                 sample_shape=N)
c = x.cat
```
together with the backward model
```python
qpi = Empirical(params=tf.Variable(tf.ones([T, K]) / K))
qmu = Empirical(params=tf.Variable(tf.zeros([T, K])))
qsigmasq = Empirical(params=tf.Variable(tf.ones([T, K])))
```
with `T` the number of samples.
Running HMC,
```python
inference = ed.HMC({pi: qpi, mu: qmu, sigmasq: qsigmasq}, data={x: x_val, c: true_c})
inference.initialize(n_print=10, step_size=0.6)
```
on this fails with
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-c3145a5e9548> in <module>()
      6 
      7 inference = ed.HMC({pi: qpi, mu: qmu, sigmasq: qsigmasq}, data={x: x_val, c: true_c})
----> 8 inference.initialize(n_print=10, step_size=0.6)

/Users/tscholak/src/edward/edward/inferences/hmc.py in initialize(self, step_size, n_steps, *args, **kwargs)
     62     self.n_steps = n_steps
     63     self.scope_iter = 0  # a convenient counter for log joint calculations
---> 64     return super(HMC, self).initialize(*args, **kwargs)
     65 
     66   def build_update(self):

/Users/tscholak/src/edward/edward/inferences/monte_carlo.py in initialize(self, *args, **kwargs)
     96     self.n_accept = tf.Variable(0, trainable=False, name="n_accept")
     97     self.n_accept_over_t = self.n_accept / self.t
---> 98     self.train = self.build_update()
     99 
    100   def update(self, feed_dict=None):

/Users/tscholak/src/edward/edward/inferences/hmc.py in build_update(self)
     88     new_sample, new_r_sample = leapfrog(old_sample, old_r_sample,
     89                                         self.step_size, self._log_joint,
---> 90                                         self.n_steps)
     91 
     92     # Calculate acceptance ratio.

/Users/tscholak/src/edward/edward/inferences/hmc.py in leapfrog(z_old, r_old, step_size, log_joint, n_steps)
    159   r_new = r_old.copy()
    160 
--> 161   grad_log_joint = tf.gradients(log_joint(z_new), list(six.itervalues(z_new)))
    162   for _ in range(n_steps):
    163     for i, key in enumerate(six.iterkeys(z_new)):

/Users/tscholak/src/edward/edward/inferences/hmc.py in _log_joint(self, z_sample)
    149     for x in six.iterkeys(self.data):
    150       if isinstance(x, RandomVariable):
--> 151         x_copy = copy(x, dict_swap, scope=scope)
    152         log_joint += tf.reduce_sum(x_copy.log_prob(dict_swap[x]))
    153 

/Users/tscholak/src/edward/edward/util/random_variables.py in copy(org_instance, dict_swap, scope, replace_itself, copy_q)
    226     kwargs['name'] = new_name
    227     # Create new random variable with copied arguments.
--> 228     new_rv = type(rv)(*args, **kwargs)
    229     return new_rv
    230   elif isinstance(org_instance, tf.Tensor):

TypeError: __init__() takes at least 4 arguments (4 given)
```

See this gist for the full code: https://gist.github.com/tscholak/14a042a65eb4e390396c9dab5aefc9f7

**Comment by tscholak** [comment]
`ed.__version__` was `'1.2.4'`,
`tf.__version__` was `'1.0.1'`.

**Comment by tscholak** [comment]
I tried the following workaround, but to no avail:
```python
pi = Dirichlet(pi_alpha)
mu = Normal(0., mu_sigma, sample_shape=[K])
sigmasq = InverseGamma(sigmasq_alpha, sigmasq_beta, sample_shape=[K])
c = Categorical(
        logits=tf.tile(
            tf.reshape(ed.logit(pi), [1, K]),
            [N, 1]))
x = Normal(mu=tf.gather(mu, c),
           sigma=tf.gather(tf.sqrt(sigmasq), c))
```
With this alternative definition for the forward model (and the backward model defined as above), the `inference.initialize` call in
```python
inference = ed.HMC({pi: qpi, mu: qmu, sigmasq: qsigmasq}, data={x: x_val, c: true_c})
inference.initialize(n_print=10, step_size=0.6)
```
throws with
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-c3145a5e9548> in <module>()
      6 
      7 inference = ed.HMC({pi: qpi, mu: qmu, sigmasq: qsigmasq}, data={x: x_val, c: true_c})
----> 8 inference.initialize(n_print=10, step_size=0.6)

/Users/tscholak/src/edward/edward/inferences/hmc.py in initialize(self, step_size, n_steps, *args, **kwargs)
     62     self.n_steps = n_steps
     63     self.scope_iter = 0  # a convenient counter for log joint calculations
---> 64     return super(HMC, self).initialize(*args, **kwargs)
     65 
     66   def build_update(self):

/Users/tscholak/src/edward/edward/inferences/monte_carlo.py in initialize(self, *args, **kwargs)
     96     self.n_accept = tf.Variable(0, trainable=False, name="n_accept")
     97     self.n_accept_over_t = self.n_accept / self.t
---> 98     self.train = self.build_update()
     99 
    100   def update(self, feed_dict=None):

/Users/tscholak/src/edward/edward/inferences/hmc.py in build_update(self)
     88     new_sample, new_r_sample = leapfrog(old_sample, old_r_sample,
     89                                         self.step_size, self._log_joint,
---> 90                                         self.n_steps)
     91 
     92     # Calculate acceptance ratio.

/Users/tscholak/src/edward/edward/inferences/hmc.py in leapfrog(z_old, r_old, step_size, log_joint, n_steps)
    163     for i, key in enumerate(six.iterkeys(z_new)):
    164       z, r = z_new[key], r_new[key]
--> 165       r_new[key] = r + 0.5 * step_size * grad_log_joint[i]
    166       z_new[key] = z + step_size * r_new[key]
    167 

TypeError: unsupported operand type(s) for *: 'float' and 'IndexedSlices'
```
Note that while this is also a `TypeError`, it's a different one than the one thrown above. This one is happening in the `leapfrog` method while the above one is in `copy`.

**Comment by dustinvtran** [comment]
Fixed in #658.

The other note seems due to the model's `tf.gather` producing a `IndexedSlices` object. I'll re-raise this as a separate issue.
